### PR TITLE
fix(metrics): Throw error when wildcard is used in search for metrics

### LINKED ISSUE #46747: [BUG] DB::Exception: Illegal type UInt64 of argument of function match
DB::Exception: Illegal type UInt64 of argument of function match

This is not the first time we incur in this issue, it has to do with the usage of match in a query that uses RELEASE_HEALTH and not PERFORMANCE. The problem happens because the match works with strings but if we are with RELEASE_HEALTH we use int-based tags due to the indexer.

from slack  

https://sentry.sentry.io/issues/3913914887/?project=1&referrer=slack

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR adds guarding to the organization metric data endpoint which will disallow the use of wildcard searches (e.g., `!release:"0.99.0 (*)"`).

Closes: https://github.com/getsentry/sentry/issues/46747

### CODE REVIEW

File: tests/sentry/api/endpoints/test_organization_metric_data.py [hunk 1]
```diff
@@ -442,6 +442,25 @@ def test_pagination_limit_without_orderby(self):
...
+            groupBy="release",
+            environment="Release",
+            query='!release:"0.99.0 (*)"',
+            statsPeriod="14d",
+            interval="1h",
+            includeTotals="1",
+            includeSeries="0",
+        )
+
+        assert response.status_code == 400
```
Comment (andriisoldatenko) [inline tests/sentry/api/endpoints/test_organization_metric_data.py:462]:
```suggestion
        assert response.body == "YOU EXPECTED ERROR MESSAGE"
```

Comment (andriisoldatenko) [inline tests/sentry/api/endpoints/test_organization_metric_data.py:462]: so check both status code and body.

**Review by andriisoldatenko — APPROVED** [review]

File: src/sentry/search/events/builder/discover.py [hunk 2]
```diff
@@ -1555,6 +1556,29 @@ def resolve_query(
         pass
 
 
+class ReleaseHealthQueryBuilder(UnresolvedQuery):
```
Comment (wmak) [inline src/sentry/search/events/builder/discover.py:1559]: Please put this either in builder/release_health.py or just put it directly in src/sentry/snuba/metrics/query_builder.py

**Comment by wmak** [comment]
> my only doubt with this PR is whether or not the user will be able to use wildcard search for non-tag values. Is that possible? I don't remember to be honest.

Neither do I, can we just test it to see if it works or not?

**Pull request merged**